nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-file-authorizer/src/main/java/org/apache/nifi/authorization/FileAuthorizer.java

```diff
@@ -0,0 +1,278 @@
+import org.apache.nifi.util.file.monitor.SynchronousFileWatcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xml.sax.SAXException;
+
+import javax.xml.XMLConstants;
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBElement;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Unmarshaller;
+import javax.xml.transform.stream.StreamSource;
+import javax.xml.validation.Schema;
+import javax.xml.validation.SchemaFactory;
+import java.io.File;
+import java.io.IOException;
+import java.util.Date;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Provides identity checks and grants authorities.
+ */
+public class FileAuthorizer implements Authorizer {
+
+    private static final Logger logger = LoggerFactory.getLogger(FileAuthorizer.class);
+    private static final String READ_CODE = "R";
+    private static final String WRITE_CODE = "W";
+    private static final String USERS_XSD = "/authorizations.xsd";
+    private static final String JAXB_GENERATED_PATH = "org.apache.nifi.authorization.generated";
+    private static final JAXBContext JAXB_CONTEXT = initializeJaxbContext();
+
+    /**
+     * Load the JAXBContext.
+     */
+    private static JAXBContext initializeJaxbContext() {
+        try {
+            return JAXBContext.newInstance(JAXB_GENERATED_PATH, FileAuthorizer.class.getClassLoader());
+        } catch (JAXBException e) {
+            throw new RuntimeException("Unable to create JAXBContext.");
+        }
+    }
+
+    private NiFiProperties properties;
+    private File authorizationsFile;
+    private File restoreAuthorizationsFile;
+    private SynchronousFileWatcher fileWatcher;
+    private ScheduledExecutorService fileWatcherExecutorService;
+
+    private final AtomicReference<Map<String, Map<String, Set<RequestAction>>>> authorizations = new AtomicReference<>();
+
+    @Override
+    public void initialize(final AuthorizerInitializationContext initializationContext) throws ProviderCreationException {
+    }
+
+    @Override
+    public void onConfigured(final AuthorizerConfigurationContext configurationContext) throws ProviderCreationException {
+        try {
+            final PropertyValue authorizationsPath = configurationContext.getProperty("Authorizations File");
+            if (authorizationsPath == null || StringUtils.isBlank(authorizationsPath.getValue())) {
+                throw new ProviderCreationException("The authorized users file must be specified.");
+            }
+
+            authorizationsFile = new File(authorizationsPath.getValue());
+            final File authorizationsFileDirectory = authorizationsFile.getParentFile();
```

Comment:
If authorizationsFile is set to something like "myFile.txt" without a directory, calling getParentFile() will return null. This would then cause a NullPointerException below. We should probably instead use:
{code}
final File authorizationsFileDirectory = authorizationsFile.getAbsoluteFile().getParentFile();
{code}